staging/src/k8s.io/api/flowcontrol/v1alpha1/types.go

```diff
@@ -0,0 +1,414 @@
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PriorityLevelConfigurationList is a list of PriorityLevelConfiguration objects.
+type PriorityLevelConfigurationList struct {
+	metav1.TypeMeta `json:",inline"`
+	// `metadata` is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// `items` is a list of request-priorities.
+	Items []PriorityLevelConfiguration `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// PriorityLevelConfigurationSpec is specification of a priority level
+// +union
+type PriorityLevelConfigurationSpec struct {
+	// `queuingType` indicates whether this priority level does
+	// queuing or is exempt.  Valid values are "Queuing" and "Exempt",
+	// and this field defaults to "Queuing".  "Exempt" means that
+	// requests of this priority level are not subject to concurrency
+	// limits (and thus are never queued) and do not detract from the
+	// concurrency available for non-exempt requests.
+	// Required.
+	// +unionDiscriminator
+	QueuingType PriorityLevelQueueingType `json:"queuingType" protobuf:"varint,1,opt,name=queuingType"`
+
+	// `queuing` holds the configuration parameters that are
+	// only meaningful for a priority level that does queuing (i.e.,
+	// is not exempt).  This field must be non-empty if and only if
+	// `queuingType` is `"Queuing"`.
+	// +optional
+	QueuingConfig *QueuingConfiguration `json:"queuing,omitempty" protobuf:"bytes,2,opt,name=queuing"`
+
+	// `exempt` holds the configuration parameters for exempt priority levels.
+	// +optional
+	Exempt *ExemptConfiguration `json:"exempt,omitempty" protobuf:"bytes,3,opt,name=exempt"`
+}
+
+// PriorityLevelQueueingType identifies the queuing nature of a priority level
+type PriorityLevelQueueingType string
+
+// PriorityLevelQueuingTypeQueueing is the PriorityLevelQueueingType for priority levels that queue
+const PriorityLevelQueuingTypeQueueing PriorityLevelQueueingType = "Queuing"
+
+// PriorityLevelQueuingTypeExempt is the PriorityLevelQueueingType for priority levels that are exempt from concurrency controls
+const PriorityLevelQueuingTypeExempt PriorityLevelQueueingType = "Exempt"
+
+// QueuingConfiguration holds the configuration parameters that are specific to a priority level that is subject to concurrency controls
+type QueuingConfiguration struct {
+	// `assuredConcurrencyShares` (ACS) must be a positive number. The
+	// API server's concurrency limit (SCL) is divided among the
+	// concurrency-controlled priority levels in proportion to their
+	// assured concurrency shares. This produces the assured
+	// concurrency value (ACV) for each such priority level:
+	//
+	//             ACV(l) = ceil( SCL * ACS(l) / ( sum[priority levels k] ACS(k) ) )
+	//
```

Comment:
I would add something communicating:
* bigger numbers mean more *reserved* concurrent requests (at the expense of every other PL).
* unused concurrent requests are returned for use by PLs which have work to do.

Reply:
The math says the first of those two things, but I suppose there is no harm in explicitly calling it out --- as long as the wording makes it clear that this is just one implication of the math.

The second thing you said is explicitly not true.  This is a big deal.  It is a big simplification that we agreed on for the first release.  I do not like it, but we discussed it and agreed that the first release would be simple in this way.

Reply:
Sorry, yes, I keep forgetting--this is an excellent place to remind future me.

Reply:
appended the first to the comment:

> 	// bigger numbers of ACS mean more reserved concurrent requests (at the
	// expense of every other PL).